modin/pandas/test/dataframe/test_indexing.py

```diff
@@ -1727,6 +1728,20 @@ def test___setitem__unhashable_list():
     df_equals(modin_df, pandas_df)
 
 
+def test___setitem__assigning_single_categorical_sets_correct_dtypes():
+    # This test case comes from
+    # https://github.com/modin-project/modin/issues/3895
+    modin_df = pd.DataFrame({"categories": ["A"]})
+    modin_df["categories"] = pd.Categorical(["A"])
+    pandas_df = pandas.DataFrame({"categories": ["A"]})
+    pandas_df["categories"] = pandas.Categorical(["A"])
+    assert_frame_equal(
+        modin_df._to_pandas(),
+        pandas_df,
+        check_dtype=True,
+    )
```

Comment:
why not to use a `df_equals`? It checks the categories equality as well, so using it should be fine.
https://github.com/modin-project/modin/blob/fea20642ac30104fe2bc4cd7b616e756f2cfac23/modin/pandas/test/utils.py#L559-L569

Reply:
`df_equals` doesn't check datatypes because it sets `check_dtype` to False. If I use `df_equals` in this test case, it passes even without my bug fix.

https://github.com/modin-project/modin/pull/3805 is an attempt to make `df_equals` always check types as well.

Reply:
[`df_categories_equal` call inside `df_equals`](https://github.com/modin-project/modin/blob/fea20642ac30104fe2bc4cd7b616e756f2cfac23/modin/pandas/test/utils.py#L569) actually should verify the categories equality, but it doesn't do so because of a bug in `df_categories_equal`. The bug is that the function doesn't verify that the amount of categorical columns at both frames is the same:
https://github.com/modin-project/modin/blob/fea20642ac30104fe2bc4cd7b616e756f2cfac23/modin/pandas/test/utils.py#L477-L483
We should rewrite these lines into something like:
```python
...
    df1_cat_cols = df1.select_dtypes(include="category").columns
    df2_cat_cols = df2.select_dtypes(include="category").columns
    assert df1_cat_cols.equals(df2_cat_cols)
    for column in df1_cat_cols:
...
```
After that verifying the categorical frames equality via `df_equals` should work fine.